### ISSUE TITLE
Incorrect error when assigning to omitted and extended record type

### ISSUE THREAD

**Issue by WearyMonkey** [Microsoft/TypeScript#27883]
**TypeScript Version:**  
typescript@3.2.0-dev.20181011
typescript@3.1.3
typescript@3.0.3

**Search Terms:**

**Code**

```ts
type Baa = { a: number, b: number };
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

function foo<P extends Baa>() {
    const foo: { foo: string } & Omit<P, 'b'> = null as any;
    const props: Baa = foo;
}
```

**Expected behavior:**

No error.

**Actual behavior:**

```
test.ts:6:11 - error TS2322: Type '{ foo: string; } & Pick<P, Exclude<keyof P, "b">>' is not assignable to type 'Baa'.
  Property 'a' is missing in type '{ foo: string; } & Pick<P, Exclude<keyof P, "b">>'.

6     const props: Baa = foo;
            ~~~~~
```

Removing the `{ foo: string } &` works.
Removing the `Omit` works.
Removing the `P` and using `Baa` directly works.

**Playground Link:** 

https://agentcooper.github.io/typescript-play/?ts=2.9.1#code/C4TwDgpgBAQghnKBeKBvKcBcUB2BXAWwCMIAnAGiiIH5t9iyoBfAbgChRIoB5AgS2AAeACqUA0gD5kUAAp8AxgGsRlAKIAPeQBs8AEwiDFEEAHsAZlFFRJE9mzN4c84HxM4oZkycEyoEdcAQOLoAzrAIEgAUAJRobFAJUPJuIcAeXtjonibYqaR8OADmzFAAZDz8QjKUAORENVIo+FpaGGFwOCDsiUkpaWCkJmAh2PCIKNnsTGxAA

**Related Issues:**

**Comment by andy-ms** [comment]
This isn't just a problem with the error message, but that we don't recognize `foo` as having any properties at all.
```ts
function f<T extends Record<"a" | "b", any>>(x: Record<Exclude<keyof T, 'b'>, any>): { a: any } { return x; }
```

`T`'s keys are *at least* `"a" | "b"`, so `Exclude<keyof T, "b">` contains *at least* `"a"`, so `x` should have `a` as a key. I think the type system currently does not have a notion on a lower bound on a type -- we can specify that `T extends something`, but not `T super something`. So we can't understand that `keyof T` will contain `"a"` even when `"b"` is excluded.

**Comment by ahejlsberg** [comment]
Fixed by #29121.
